Array index analysis using SMT

mn416 opened the pull request:
This PR introduces an `ArrayIndexAnalysis` class to determine whether or not distinct iterations of a given loop can generate conflicting array accesses (if not, the loop can potentially be parallelised). It formulates the problem as a set of SMT constraints over array indices which are then are passed to the third-party [Z3 solver](https://github.com/Z3Prover/z3).

The PR also introduces a new option `use_smt_array_index_analysis` to `DependencyTools`, which enables use of the new analysis in place of `DependencyTools`'s current array conflict analysis. Transformations such as `ParallelLoopTrans` and `OMPLoopTrans` are parameterised by `DependencyTools`, allowing use of the new analysis in existing transformations.

To install the Z3 solver:

```
$ pip install z3-solver
```

This dependency has been added to `setup.py`.

The analysis is broadly explained in [source code comments](https://github.com/stfc/PSyclone/blob/mn416-smt-array-index-analysis/src/psyclone/psyir/tools/array_index_analysis.py).

There is also a set of examples showing what the analysis is capable of, available [here](https://github.com/mn416/array-analysis-examples). None of these examples are solved by PSyclone's existing loop analysis, and all can be solved by the SMT-based analysis.

The comments below include results from applying the analysis to the full UKCA codebase (execution time of the analysis, and number of loops parallelised).

mn416 committed:
New SMT-based ArrayIndexAnalysis

mn416 committed:
Support arbitrary precision integers in addition to bit vectors

Also use z3-solver instead of pysmt as it exposes additional useful
features, such as predicates for checking bit vector overflow.

mn416 committed:
Add more SMT-related options to ParallelLoopTrans

mn416 committed:
New analysis option to prohibit bit vector overflow

arporter commented:
This feels like one for @hiker to look at as he did DependencyTools - without looking at the code my feeling is that it would make sense for extra sophistication to go in there rather than into a specific PSyIR node.

EDIT: my mistake - I see you say you've added a new class to encapsulate this functionality. I'd still like to see whether it makes sense to merge it/include it with DependencyTools though.

mn416 committed:
Add z3-solver to setup.py

mn416 committed:
Handle structures correctly

mn416 committed:
Call ArrayIndexAnalysis from DependencyTools instead of ParallelLoopTrans

mn416 committed:
Simple heuristic to decide between integer and bit-vector solvers

mn416 committed:
Small tweak to heuristic

mn416 committed:
Option to handle some common array intrinsics

mn416 committed:
Improve testing and coverage

mn416 committed:
Make flake8 happy again

mn416 committed:
Add a test for a failed analysis

mn416 committed:
Move some tests to dependency_tools_test.py for coverage

mn416 committed:
Add more pydoc and tests

mn416 committed:
Use a clearer encoding for loop variable constraints

mn416 committed:
flake8

mn416 committed:
Add solver sweeper to reduce solver sensitivity to constraint order

mn416 committed:
Add missing return type

mn416 committed:
Performance improvements

mn416 committed:
Remove unnecessary clause from IfBlock.flat()

mn416 committed:
Make the solver timeout optional

hiker commented on src/psyclone/psyir/nodes/if_block.py:
```diff
@@ -195,3 +196,22 @@ def reference_accesses(self) -> VariablesAccessMap:
+           (if there are any) are recursively gathered. The condition for
+           the final 'else' in the chain (if there is one) is 'None'.
+        '''
+        if self.else_body is None:
```
I find this harder to understand than necessary (probably because adding 'this' block's data at the end with an `insert`). As far as I can tell, the following code produces the same result:
```
        branches = [(self.condition, self.if_body)]
        if self.else_body:
            if (isinstance(self.else_body, Schedule) and
                    len(self.else_body.children) == 1 and
                    isinstance(self.else_body.children[0], IfBlock)):
                branches.extend(self.else_body.children[0].flat())
            else:
                branches.append((None, self.else_body))
        return branches
```
(if not, we need a test for that, the code passes the existing tests).

hiker commented on src/psyclone/tests/psyir/nodes/if_block_test.py:
```diff
@@ -282,3 +283,72 @@ def test_ifblock_children_validation():
         ifblock.addchild(else_body)
     assert ("Item 'Schedule' can't be child 3 of 'If'. The valid format is: "
             "'DataNode, Schedule [, Schedule]'." in str(excinfo.value))
+
```
Can you also add a test with no `else` branch? ATM that line in `if_block.py` is not covered. My suggested re-writing avoids this issue (since there is no code to execute if there is no else block :) ), but still we should have a separate test that covers the no-else case.

hiker commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -884,9 +896,15 @@ def can_loop_be_parallelised(self, loop,
-                                                             var_info)
+                # If using the SMT-based array index analysis then do
+                # nothing for now. This analysis is run after the loop.
+                if self._use_smt_array_index_analysis:
```
Would it make sense to first run the existing array analysis? And only run the the Z3 solver if the result is false (and long term, maybe the existing analysis could return three values: yes, no, unknown.

I might try to do some additional performance tests (I noticed the results you have posted ;) ), this obviously only makes sense if the existing one is significantly faster.

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,1099 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2025, University of Cambridge, UK.
```
I need to check with Andy, but I believe this should have the STFC header :(

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,1099 @@
+# The analysis class provides a method 'is_loop_conflict_free()' to
+# decide whether or not the array accesses in a given loop are
+# conflicting between iterations. Two array accesses are conflicting
+# if they access the same element of the same array, and at least one
```
... the same array in different loop iterations, ...

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,1099 @@
+#
+# The analysis assumes that any scalar integer or scalar logical
+# variables written by the loop can safely be considered as private
+# to each iteration. This should be validated by the callee.
```
callee and is typically done in the DependencyTools.

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,1099 @@
+# not be executed at run time.
+#
+# We also maintain a "current condition".  This can be viewed as a
+# constraint that has not been comitted to the constraint set because
```
s/comitted/committed/

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,1099 @@
+# exit conditional blocks of code. This current condition is passed in
+# recursive calls, so there is an implicit stack of them.
+#
+# More concretely, when we encouter an 'if' statement, we copy the
```
s/encouter/encounter/

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,1099 @@
+# written by the loop body, because we don't know whether we are
+# entering the loop (at run time) for the first time or not. Second,
+# we create two SMT variables to represent two arbitary but distinct
+# iterations of the loop. Each variable is constrained to the start,
```
Each of these two variables is constrained to ...

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,1099 @@
+# stop, and step of the loop, and the two variables are constrained to
+# be not equal. After that, we analyse the loop body twice, each time
+# mapping the loop variable in the substitution to each of the SMT
+# loop variables.  After analysing the loop body for the first time,
```
Single space between `.` and `After` (that happened elsewhere here as well, maybe just search for `.  ` (double space)

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,1099 @@
+# the loop.
+#
+# When the recursive descent is complete, we are left with two array
+# access lists. We are interested in whether any pair of accesses to
```
What about replacing 'We are interested ...  conflict' with something like:

A conflict happens if there is an access in the first list of accesses that is the same as an access in the second list, and one of which is a write.

hiker replied:
... access lists, representing two different iterations of the same loop.

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,1099 @@
+        self.opts = options
+
+    def _init_analysis(self):
+        '''Intialise the analysis by setting all the internal state
```
s/Intialise/Initialise/

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,1099 @@
+
+    def _init_analysis(self):
+        '''Intialise the analysis by setting all the internal state
+        varibles accordingly.'''
```
s/varibles/variables/

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,1099 @@
+                           ) -> (z3.ExprRef, z3.BoolRef):
+    '''Translate a scalar integer Fortran expression to SMT. In addition,
+    return a constraint that prohibits/ignores bit-vector overflow in the
+    expression.'''
```
Could you add a `:returns:... ` line

hiker reviewed (commented):
I don't have enough time today to finish the review, so I have to abort somehow in the middle.
Two 'general' questions:
1. Is there any advantage of supporting bit vectors? It feels like the transformation of expressions to Z3 can be cut in half if we would only support integers (and since we kind of agree to ignore overflows).  Or of course the reverse ;)
2. I feel it might be useful to have a command line option to generically enable your analysis and set the main configurations (timeout and #threads?). E.g. the Fab build system should be able to support to directory-specific flags, so the build system could enable your analysis for all files under `../ukca` or so, and not affecting anything else?

That might be useful if the same transmutation script is used across different source directories, but only a subset will benefit from Z3.

I will continue hopefully tomorrow.

hiker commented:
...
> Here is a histogram of the times taken by each method to complete on each loop:

Do you have the file names for the slow ones? I would be interested to look at them in more detail (for both approaches).

...
> As expected, `DependencyTools` is generally quicker. Its run time is proportional to the size of the loop whereas the run time of `ArrayIndexAnalysis` is proportional to the size of the routine containing the loop. Plus there is the cost of running a general-purpose solver.

I assume you are looking at the code 'around' a loop to know additional constraints on variables that might be used in the loop? Is this actually required (i. e. leads to more parallelisable loops)? I still need to fully understand you code ;)

Might an option to restrict the Z3 analysis to just the loops be useful?? 


> All loops taking longer than 15s to analyse by `ArrayIndexAnalysis` contain auto-generated code (unnatural loops containing hundreds/thousands of array access sites). 

Would that imply that you get hundreds of duplicated constraints (which could slow things down)? Maybe this could be filtered??


> I have not investigated the `DependencyTools` outliers.
> 
> The `ArrayIndexAnalysis` was run with an SMT-solver timeout of 5s, and none of the loops in UKCA timed out. This is an encouraging sign that the SMT solver is generally good at solving array-conflict problems.

Sounds promising.
> 
> Using `ArrayIndexAnalysis` led to 12 exceptions. These were all due to Fortran code not supported by PSyclone (referring to the same array multiple times on the LHS of an assignment).
> 
> Using `DependencyTools` led to 54 exceptions. I have not yet explored the cause of these.

Again, I would be interested in a list :)

If you don't have the data ready, don't worry, I should be able to reproduce them:
```
science/ukca/src$ find  -iname "*.f90"|wc
    308     308   15702
```
which seems to match the number of files you mentioned :)

hiker commented:
I just realised that I am running on a slightly different ukca version, but here outliers  (runtime > 5 seconds) I see: (from 256 files, after which psyclone appears to hang, which I will look at separately):
```
File name                   loop#                           DepTools    parallel?   z3        z3 result

ukca_um_strat_photol_mod.F90	2	Execution time loop i:	11.899324	True	1.511083	True
ukca_um_strat_photol_mod.F90	3	Execution time loop j:	6.690157	True	1.471593	True
ukca_chemistry_ctl_col_mod.F90	2	Execution time loop j:	0.032098	False	5.312659	False
ukca_deriv_aero.F90				1	Execution time loop n:	0.173906	False	16.631266	False
ukca_deriv_aero.F90				4	Execution time loop i:	0.331547	False	14.088031	False
ukca_deriv_raqaero_mod.F90		1	Execution time loop n:	0.379767	False	25.602987	False
ukca_deriv_raqaero_mod.F90		3	Execution time loop i:	0.126291	False	24.398631	False
ukca_surfddr.F90			   17	Execution time loop j:	0.196217	False	8.116066	False
ukca_aerod.F90					1	Execution time loop j:	38.493939	True	5.474127	True
ukca_deriv_raq.F90				1	Execution time loop n:	0.291267	False	20.539170	False
ukca_deriv_raq.F90				3	Execution time loop i:	0.110052	False	21.570237	False
ukca_deriv.F90					1	Execution time loop n:	0.268167	False	11.196480	False
ukca_deriv.F90					3	Execution time loop i:	0.109296	False	10.146189	False

```

hiker commented:
I will stop for now, and hand it back to you. I started to get into the actual analysis, but I thinks splitting this into two files (one to convert expressions (or Fortran statement), one to do the actual analysis) will be a good start.

Note that I also did #3311, which should fix some (all?) of the timing outliers with the dependency tools ;) Thanks for your performance results.

mn416 committed:
Adjust `IfBlock.flat()` as suggested by @hiker

Also add a test for a chain with no terminating 'else'.

mn416 committed:
Simplify `use_smt_array_index_analysis` parameter

mn416 committed:
Generate conflict messages in ArrayIndexAnalysis

mn416 committed:
Modify ParallelLoopTrans to take an optional DependencyTools instance

mn416 committed:
Ensure that `test_array_analysis_option()` introduces OpenMP directives

mn416 committed:
Tidy ArrayIndexAnalysis comments/docs

mn416 committed:
Tidy comments/docs in `translate_expr()`

mn416 committed:
Add getters for the ArrayAccess class

mn416 committed:
Drop 'type_ok' checks in SMT expression translation

mn416 committed:
More comments/docs in ArrayIndexAnalysis

mn416 committed:
Don't access private '_signature' field directly

mn416 committed:
Recognise `stop` and `error stop` statements

mn416 committed:
Split array index analysis and Fortran-to-Z3 translation

mn416 committed:
Provide a default argument in `FortranToZ3.solve()`

mn416 committed:
flake8

mn416 committed:
Merge branch 'master' into mn416-smt-array-index-analysis

mn416 committed:
Avoid python >= 3.10 features

mn416 committed:
Count solver timeouts as test passes

Also reduce the timeout to avoid possibility of unduly slowing down the
test suite.

mn416 committed:
Add "no cover" pragma for difficult to reach statement

mn416 committed:
Add docs to the developer guide

mn416 commented:
@hiker, thank you very much for all your comments and suggestions. I think I've addressed all of them, except for a few outstanding ones which I address below.

First, and most importantly, you had a concern about the use of a solver timeout. Paraphrasing:

> On a slow (or highly loaded) machine, the time out could be reached and a loop would not parallelised, but on a faster (not highly loaded) machine, the loop would be parallelised. That's ... really not something we should be doing.

This is a valid concern and, in fact, I experienced it when the PSyclone CI started running my tests. As a result, I modified my tests so that timeouts are not treated as failures. However, it does bring into question whether the analysis should be used in transmutation scripts where the user depends on a consistent result. I added the following warning to the developer guide:

> The main drawback of the SMT-based analysis is that the time-to-solve is not guaranteed to be consistent. It might be fast using one version of Z3 but slow using another. It might vary significantly with seemingly innocuous code changes. Having said that, our experience is that it generally performs well, especially when setting the option `num_sweep_threads=4` to explore different constraint orderings in parallel (this is the default setting). Nevertheless, it is probably best suited to static checking / linting, e.g. spotting bugs in parallel loops or missed opportunities for parallel loops. It is probably not advisable to use it in cases where a consistent result is relied upon, such as a transmutation script.

I will leave it to you to decide whether or not it is a good idea to go ahead with this PR. Obviously, please do what is right for the PSyclone project (I will not be dissappointed or offended). I think the analysis has some great potential uses in a static checker, but that may not align with the goals of the PSyclone project.

Elsewhere, you suggested splitting out the translation-to-Z3 code from the analysis. This has been done in commit dc0bba0. The code has also been brought up-to-date with master.

Things I can look at further if you're happy to go ahead with the PR:

  * I've added documentation to the developers guide (commit 1bb4705), but not done the `argpase` suggestion and associated changes to user guide.

  * Your suggestion to restrict the analysis to just the loops (rather than looking at the whole routine) probably would help with analysis time and might work well as a new option, but I've not tried it.

BTW, all CI checks were passing before the I made some documentation changes, but now code coverage is failing even though all new lines of have been covered.

mn416 committed:
Add `succeed_on_timeout` option

hiker commented on src/psyclone/psyir/tools/fortran_to_z3.py:
```diff
@@ -0,0 +1,536 @@
+        elif call.intrinsic == IntrinsicCall.Intrinsic.UBOUND:
+            var = "#ubound"
+        else:
+            return None
```
What are the implications if an unsupported 'inquiry` function is called? It is de-facto ignored? I am not 100% certain about the potential impact, but maybe we need a `unsupported` section in the manual. Also, if this could result in incorrect results, should we return that this is not parallelisable?

hiker commented on src/psyclone/psyir/tools/fortran_to_z3.py:
```diff
@@ -0,0 +1,536 @@
+                var = var + "_" + sig.var_name
+                if len(call.children) == 3:
+                    rank = call.children[2]
+                    if isinstance(rank, Literal):
```
Please comment on what is happening here.

hiker commented on src/psyclone/tests/psyir/tools/array_index_analysis_test.py:
```diff
@@ -0,0 +1,381 @@
+
+          n = size(arr)
+          do chunk_begin = 1, n, chunk_size
+            chunk_end = min(chunk_begin+chunk_size-1, n)
```
Again, add another example that has an overlap and is detected.

hiker commented on doc/developer_guide/dependency.rst:
```diff
@@ -585,6 +585,34 @@ can be parallelised:
+as in a build-time transmutation script, it should be used with the
+`succeed_on_timeout=True` option. This means the analysis may report no
+conflicts when there is one, but it will have tried to find one.
+
```
Can you also add documentation (e.g. autodoc) about the `ArrayIndexAnalysisOptions`, since this is being used by developers.

hiker reviewed (commented):
A bit more coming, just making sure what I have so far is not lost

hiker commented on src/psyclone/psyir/tools/array_index_analysis.py:
```diff
@@ -0,0 +1,827 @@
+           :param loop: loop to be analysed.
+           :param all_conflicts: if True, enumerate all conflicts, otherwise
+              stop after the first conflict. Defaults to False.
+           :return: a list of pairs array-name/message pairs. If the list
```
Can you add an empty line between the parameter and `:return:`? I seem to remember that it was required in the past (otherwise sphinx would fail). I tried, and in this case it seems to work. Not sure if I have a specific newer/older version installed. Since for now it's coding style, let's play it safe and add a new line here.

hiker reviewed (changes requested):
Thanks a lot for the work here, good stuff. I have a few issues to be addressed, but we are making progress - thanks!
